kge/model/transr.py

```diff
@@ -0,0 +1,121 @@
+            # n = n_s = n_p = n_o
+            s_emb_translated = self._transfer(s_emb, rel_emb, projection_matrix)
+            o_emb_translated = self._transfer(o_emb, rel_emb, projection_matrix)
+
+            out = -f.pairwise_distance(s_emb_translated + rel_emb, o_emb_translated, p=self._norm)
+        elif combine == "sp_":
+            # n = n_s = n_p != n_o = m
+            m = o_emb.size(0)
+
+            s_emb_translated = self._transfer(s_emb, rel_emb, projection_matrix)
+            s_emb_translated = s_emb_translated.repeat(m, 1)
+
+            rel_emb = rel_emb.repeat(m, 1)
+            projection_matrix = projection_matrix.repeat(m, 1)
+
+            o_emb = o_emb.repeat_interleave(n, 0)
+            o_emb_translated = self._transfer(o_emb, rel_emb, projection_matrix)
+
+            out = -f.pairwise_distance(s_emb_translated + rel_emb, o_emb_translated, p=self._norm)
+            out = out.view(m, n).transpose(0, 1)
+        elif combine == "_po":
+            # m = n_s != n_p = n_o = n
+            m = s_emb.size(0)
+
+            o_emb_translated = self._transfer(o_emb, rel_emb, projection_matrix)
+            o_emb_translated = o_emb_translated.repeat(m, 1)
+
+            rel_emb = rel_emb.repeat(m, 1)
+            projection_matrix = projection_matrix.repeat(m, 1)
+
+            s_emb = s_emb.repeat_interleave(n, 0)
+            s_emb_translated = self._transfer(s_emb, rel_emb, projection_matrix)
+
+            out = -f.pairwise_distance(o_emb_translated - rel_emb, s_emb_translated, p=self._norm)
+            out = out.view(m, n).transpose(0, 1)
+        else:
+            return super().score_emb(s_emb, p_emb, o_emb, combine)
+
+        return out.view(n, -1)
+
+
+class TransR(KgeModel):
+    r"""Implementation of the TransR KGE model."""
+
+    def __init__(
+            self,
+            config: Config,
+            dataset: Dataset,
+            configuration_key=None,
+            init_for_load_only=False,
+    ):
+        self._init_configuration(config, configuration_key)
+
+        transr_set_relation_embedder_dim(
+            config, dataset, self.configuration_key + ".relation_embedder"
+        )
+
+        super().__init__(
+            config=config,
+            dataset=dataset,
+            scorer=TransRScorer,
+            configuration_key=self.configuration_key,
+            init_for_load_only=init_for_load_only,
+        )
+
+
+def transr_set_relation_embedder_dim(config, dataset, rel_emb_conf_key):
+    """
+    Set the relation embedder dimensionality for TransR in the config.
+    """
+    ent_emb_conf_key = rel_emb_conf_key.replace(
+        "relation_embedder", "entity_embedder"
+    )
+    if ent_emb_conf_key == rel_emb_conf_key:
+        raise ValueError(
+            "Cannot determine relation embedding size. "
+            "Please set it manually to 'k * (1 + d)'. "
+            "For d = entity embedder dimensionality. "
+            "For k = relation embedder dimensionality. "
+        )
```

Comment:
Is this part necessary? 
dim_d is from entity_embedder.dim and dim_k is relation_embedder.dim 
and new_dim_k is calculated by dim_d and dim_k. 
So I don't think that we need to check this.

Reply:
I'm not sure whether it is necessary (I guess not) but since TransH had this check I added it: https://github.com/LJ171/kge/blob/d0634774b1b29afc739454737aa105dec6cb60b2/kge/model/transh.py#L156
Should I remove it?

Reply:
This part is used in RESCAL and TransH, and the reason is their default relation embedding dimenstion is -1 which is to determine relation dimension automatically.
But in TransR, the relation embedding dimenstion can't be determined automatically (so default value is not -1), so I think this part is unnecessary.

Reply:
Ahh, okay make sense, then I will remove it.

Reply:
Addressed with https://github.com/LJ171/kge/pull/2/commits/f3ca3017653f7509a38c643c12e0c24840fe278d.